### DISCUSSION TITLE
Subscription plugin

### DISCUSSION THREAD

**Discussion by rrums123** [medusajs/medusa#1232] [Q&A]
Hi all!

I need a subscription feature with Medusa, so I think I need to build a plugin that can do the following.

- Give options to user for a one-time purchase or time-based subscription (weekly or monthly)
- Adding a digital product variant, because the subscription feature is very related to this type of product. This type of product does not require a shipping feature, so it can be skipped.
- Automatic invoicing will be sent every time period (weekly or monthly)
- Payment using Stripe

To realize the above idea, I think I need to implement the following things

- Add digital product type. From the existing database schema, I think there is no need to create a new table, but by adding record to the `product_option`, `product_option_value` and `product_variant` table everytime digital product type is selected.
- Modify how orders are placed, if everything in the cart is digital, shipping is skipped.
- Use the payment plugins (Stripe, etc) to subscribe, then create cron jobs to generate invoices automatically

If there is any input on how to implement it better or technical implementation that I need to explore first, I really appreciate it. Please consider that I am very junior to node.js.

Thanks for your time.

**Comment by ivanminutillo** [comment]
Probably out of scope for your use case, but I was looking for a subscription feature as well that allows me to use medusajs also as something similar to a replacement for Ghost (the blogging platform) subscription feature. 
So basically I  have a set of digital contents (a bunch of blog posts, podcasts, etc.) and I want to add a paywall on some of them, so that they can be experienced only if the user is subscribed, meanwhile the rest is open.
Maybe I want to push the boundaries of the framework toward things it is not meant to be used, I am new to medusa and im mostly trying to see if I can use the same tool for different use cases i will work on soon.

**Comment by spenya** [comment]
Hi, we are also looking for a subscription feature for our development ... we will have members that will pay monthly, anually or maybe for 6 months, so for diferents periodicities would be important ... We are not very used to develop in node neither but we will be happy to help if we can! thanks

**Comment by adrien2p** [comment]
Some data inputs,

[Here](https://github.com/stripe-samples/subscription-use-cases/blob/master/fixed-price-subscriptions/server/node/server.js) is an example of subscription payment that can be setup using stripe. That could be an update to the stripe plugin. Here is the [doc](https://stripe.com/docs/billing/subscriptions/overview)
And here the [Api reference](https://stripe.com/docs/api/subscriptions/object).

Each product from the cart can be part of different subscription type?

If the order has classic products and subscription products there should be a one time payment for the classic products and a subscription for the other products. It must also be reflected into the view. The customer should also be able to cancel the subscription 

The stripe plugin must use the eventbus service to emit the different necessary events that the webhook could received.

**Reply by adrien2p** [reply]
As an admin I need:

Subscription:

- To create subscriptions
- Update subscriptions
- delete subscription for future purchase.
- see all active subscriptions
- see the total recurring amount

What type of subscriptions? 
- subscription to the platform to access some products
- subscription related to a product or variant?
- both

Hook:

- does the provider generate and send the invoice through a hook or should we have to get them periodically? To be able to notify the customer. Is it something the provider provide?

Product:
- create variants that required a subscription? Then select the subscription types available? One variant per subscription type?
- create a variant that start a subscription related to itself? Then being able to choose the subscription types available? One variant per subscription type? Looks like rental?

As a customer:

- allow to add or not some products based on the required subscription?
- some product could start a subscription define by some@ characteristics? Rental?
- display both one time payment amount and subscriptions amount if needed in the order summary
- being able to cancel subscription.
- if product based subscription, being able to cancel it? Is there any constraint?
- choose a plan if any

API:

Product list:

- a wrapper method could wrap the products/variants based on the current logged in customer to add some transient data to easily manage the client view on does the product/variant can be added to the cart or not and why. That way any custom admin client can also easily integrate the feature

Cart:

- final validation on the customer eligibility for the cart/order items

Probably other point I did not think about at midnight 😂

**Reply by rrums123** [reply]
> What type of subscriptions?

> subscription related to a product or variant?

Subscription related to variant, each variant can be purchased as one-time purchase or subscribe
For now we want to use this feature to subscribe to physical products, so the product will be sent to customer periodically.

> does the provider generate and send the invoice through a hook or should we have to get them periodically? To be able to notify the customer. Is it something the provider provide?

Invoice should be realeased periodically

> create variants that required a subscription? Then select the subscription types available? One variant per subscription type?


yes, just like this product in Amazon
![amazon](https://user-images.githubusercontent.com/100153954/159886760-6398244e-a1fd-4b48-95d4-686921c02d7c.png)

> display both one time payment amount and subscriptions amount if needed in the order summary

yes, that is needed

> being able to cancel subscription.

> if product based subscription, being able to cancel it? Is there any constraint?

user can cancel subscription without any constraint

**Comment by corend** [comment]
I subscribe to this topic, this is not a matter of choice in 2022, a lot of merchants use subscriptions as business model. So I also need it for my customers.

**Comment by klauss194** [comment]
We have started coding it. As a simple plugin for now. 
Anyone -> please list your requirements in this discussion.

**Reply by pepijn-vanvlaanderen** [reply]
@klauss194 cool that you are working on this! We are exploring e-commerce possibilities with Medusa and this is the only big feature that is currently missing. 

From my experience with other subscription modules in other e-commerce tools these are some basic requirements:

### Schedule
- Billing interval (every, every 2nd, every 3rd, every 4th, every 5th, every 6th)
- Billing period (day, week, month, year)
- Start date
- Trial end date
- Next payment date
- End date

### Status
- Pending (created, but payment of first order is not yet made)
- Active 
- On hold (awaiting payment)
- Cancelled
- Expired (after end date of subscription)
- Pending Cancellation (depends on subscription type, but this can be the case when you buy a yearly subscription and cancel it mid-term, so the rest of the term it is active, but will not renew) 
 
### Line items
Same as order line items. This should be editable also from admin perspective and user perspective, so the subscription can change overtime.

### Timeline
Same as order timeline, to keep track of all related subscription events.

### Payment method
Some info on the used recurring payment method. Detailed payment info should not be necessary, that can be viewed through related orders

### Related orders
A list of all the parent order (first order) and all renewal orders made for this subscription.

### Downloadable product permissions
We do not really have use-cases for this atm, but I can imagine for digital goods you want some special configurations.

### Actions
- Process renewal (manually trigger the renewal process, same as be done in cronjob when the next payment date is passed)
- When having downloadable products, regenerate permissions

**Reply by rrums123** [reply]
@pepijn-vanvlaanderen hey thanks for the insights, we will consider it for the plugin

**Comment by corend** [comment]
Hi,

Sorry for the late response. I was quite busy.

I come from the world of payment, so I allow myself to give my point of view. And what personally would help me in my dealings with my customers.

To express the need I will position myself on the frontend side (client) which is often the case for our customers, of course on the server side it must be managed for it to be operational. But let's keep a more functional view on the frontend side.

For my part, the concrete need for a subscription extension must absolutely meet the following needs:
When I am the store administrator, I must be able to configure a precise list of subscriptions which must be configurable by period. I must then have a link to a specific subscription management page in "products" allowing me to configure my subscriptions via a form with the following information:
- All current information valid for a product (name, description etc.)
- Is the subscription active or not (what about deactivation for customers who have already subscribed, for me the subscription must continue until interruption either by himself or by an admin user)
- In addition in the Variants it should only be possible to configure the following variations:
-- subscription price,
-- the renewal period (daily, weekly, monthly, every 3 months, every 6 months -or every X months to be defined-, or annually),
-- when the first direct debit must be executed (first day of the period, first day of the following month, immediately...)
-- the date when the direct debits will be taken according to the period chosen: if it is daily then choose a time of day, if it is weekly then propose a day of the week with Monday by default, for the month then choose a date in the month with the 1st of the month by default

Regarding the statuses, it must be possible to have:
- Pending status
- Active status
- Inactive Status
- Frozen
- Closed
- Full refund
- Partial refund

Also in the admin section, it should also be possible to create subscriptions for a person or a company from scratch with the same settings.
Also in the admin section, it must be possible to manage current subscriptions with a view (get) on current subscriptions and give the possibility of stopping a subscription and/or refunding it for part or all of the direct debits .

In the test store, it would be nice to be able to view these subscriptions easily as well. This would facilitate customization starting from an example store managing subscriptions.

Another type of subscription: the discount based on a subscription.
There are more and more businesses based on the business model of having raw product prices that can receive a discount based on a subscription price (usually annual).
For example, you have products in the store, and if you take an annual subscription then you can get a discount on specific products or on the entire store.

Last type of subscription that is increasingly emerging, which for me is really optional at the implementation level because it is quite rare, these are models based on a subscription rather of the recurring order type. This means that we flag one or more products that can be renewed, the customer then chooses how many items he wants to receive and for what period (every week etc...). For example, if you have a baby, you know you need X diapers every Y weeks so you can configure yourself to receive them every Y weeks until you stop this subscription loop yourself .

Obviously in a more global way, for all solutions, it must be possible for the customer to be able to manage his subscriptions via a dedicated member area: to manage his payment methods, add a subscription, cancel a subscription, freeze a subscription , resume a subscription etc.

I wonder how are you going to manage the payment method storage?

I really hope this can help you. In any case if the first type of subscription is already made, that I think would help a lot of business.

Thank you for all the work you do.

**Comment by rrums123** [comment]


**Reply by rrums123** [reply]
Updated the links, previously entered the wrong links.

**Reply by pKorsholm** [reply]
Hey @rrums123, do you mind if I add these to the awesome-medusajs repo that @adrien2p has created?

**Reply by badrange** [reply]
Would be really nice if you could share the plugin @rrums123 and if it could be added to awesome-medusajs-repo! :+1:

**Reply by badrange** [reply]
Oh it's there, I didn't see it at first, it's under the Payment headline:
https://github.com/adrien2p/awesome-medusajs#payment

**Reply by straussCS** [reply]
@rrums123 
The link does not seem to be valid anymore: https://github.com/rrums123/medusa-plugin-subscription
Is the plugin still available?

**Reply by Krislunde** [reply]
I'm also wondering about this. Link doesn't work, but am desperately trying to find a good ecommerce system with subscriptions.

**Reply by badrange** [reply]
The repo has been removed/made private.

**Reply by klauss194** [reply]
Hi, 

The plugin wasn't finsihed -> It was just an initial commit -> it was made offline because I didn't know if its wise to releasesomething that is so hardcoded. But I'd be happy to assist on discord -> klauss#0816

**Comment by david-bermejo** [comment]
Hi,

Is there any update on this? I am stuck at integrating Stripe subscriptions with current MedusaJS cart and orders implementation. The problem is that there is no way to update the cart's paymentIntent, which is created automatically by medusa's stripe plugin, with the one associated with Stripe subscription object, and hence when the payment occurs the cart cannot be marked as completed and the order not created.

Any help would be appreciated!

**Reply by greendesertsnow** [reply]
Well they offer a tutorial on how to build your own subscription based store
https://docs.medusajs.com/recipes/subscriptions

**Reply by benbucksch** [reply]
a) That requires to build all the subscription logic manually. b) It needs close integration with all (!) the payment providers used, and many other aspects are needed.

**Comment by benbucksch** [comment]
As David [mentioned](https://github.com/medusajs/medusa/discussions/1232#discussioncomment-7610678), the subscription must be integrated with the payment provider. Stripe and PayPal both have specific payment paths for subscriptions, both initial payment and renewal. That's necessary, because during a renewal charge, the user is not present to do 2-factor auth like 3D Secure / Verified by Visa, so the initial payment needs to already authorize for later charges. This is a specific payment flow, so it needs to be integrated with the payment provider.

Subscriptions also need to handle
* renewal failures, like credit card expiry, too low account funds, and dynamic re-attempt of payment without customer presence, to recover from payment failures
* email the user on renewal failure, in reasonable intervals, with the ability to enter a new card - preferably pre-authenticated.
* email the user a few weeks *before* the renewal charge to allow him to cancel before renewal (legal requirement)
* trigger the renewed provisioning of the product.

I've implemented that in a custom store in the past, and I don't want to have to do that again. This is a serious amount of work. It's far more work than the [recipe](https://docs.medusajs.com/recipes/subscriptions) makes it looks like. Testing is also not trivial.

Medusa looks great. But given that our business is entirely based on subscriptions, this is a pretty hard setback.

**Comment by minhchu** [comment]
@benbucksch thanks for your feedback

Our business offers both subscription plans and simple products. Any ideas on how to implement both flows would be greatly appreciated.

**Comment by manishkumar212111** [comment]
@rrums123 @klauss194 I am also looking for subscription/one time based purchase options. is there any plugins available for same

**Comment by klauss194** [comment]
Guys we are planning to release some of our work. @manishkumar212111 can you please detail in this thread your specifications ?

**Reply by manishkumar212111** [reply]
> Guys we are planning to release some of our work. @manishkumar212111 can you please detail in this thread your specifications ?

Subscription:

- To create subscriptions (Weekly, Daily, Monthly, Alternate days)
- see all active subscriptions
- Automatic deduct amount from Wallet once delivered
- Automatic Cancel Subscription once wallet amount is less than subscribed product amount.
- Pause/Resume Subscription

Wallet
- Recharge Wallet
- View Wallet
- View Transactions

Admin-

- Can view subscribed order
- Orders will be created for every slot delivery
- View wallet, Add Money to wallet

**Reply by klauss194** [reply]
View wallet, Add Money to wallet <- can you develop a bit on this one ? how does the wallet play in the whole ecosystem ?

**Reply by manishkumar212111** [reply]
> View wallet, Add Money to wallet <- can you develop a bit on this one ? how does the wallet play in the whole ecosystem ?

It will save us from autopay from payment gateway. For user also it will be tough for autopay. I will develop wallet plugin. once subscriptionplugin is integrated

**Reply by manishkumar212111** [reply]
@klauss194 May I know any tentative timeline for subscription plugin release

**Reply by Krislunde** [reply]
@klauss194 I would like to add on to this:
- the ability to have both single purchase items and subscription items in the cart at the same time for one single checkout. 

This is a serious expectation many clients have because Shopify offers this feature. We were seriously considering Medusa for our stack, because we love and support projects that are based on open source and self-hosting, without vendor tie-in, but the amount of work to build this into Medusa forced us to go with Swell instead (which also offers the ability for single-purchase items and subscriptions in the same cart.)

Swell also creates separate new "orders" for the retailer for each item that is part of a single subscription, instead of a single order with all the items on the same renewal schedule. This is a drawback, but granted, a minor one.

**Reply by manishkumar212111** [reply]
@klauss194 Any update on this ?

**Reply by benbucksch** [reply]
@klauss194 I'd be happy to chime in. You may contact me on com.shop at mustang dot im

**Comment by benbucksch** [comment]
@klauss194 These are our requirements for subscription:

## Features
* Subscriptions: Automatically trigger monthly/yearly renewal payments.
  * Upon subscription, make first real payment for the first subscription period. Subscription only happens, if that first payment succeeded.
  * First payment needs already needs to inform the payment provider (Stripe credit card, PayPal, etc., see below) that we will make further charges without the customer present. This is typically a specific flag to set when requesting payment.
  * Renewal payments are made without the customer present, without multi-factor auth (MasterCard 3D secure, Verified By Visa, etc.), using the customer record at the payment provider and the permission granted (see previous point).
* Payment options
  * Stripe, for credit card, WeChat and others
  * PayPal
  * SEPA direct debit (we charge the customer's EU bank account automatically)
  * Manual wire transfer by customer to us
  * Purchase Orders (POs) for purchase departments of large companies and governments

Needs to work together with other components:
* Taxes are calculated automatically
* Invoices are created automatically and sent by email to the customer
* Affiliate program

## UI
### Subscribe
Customer is on our website and wants to buy/subscribe to our products. We will build our own custom shop frontend, using REST APIs.

### Customer self-management
We need a way to allow end users and companies to:
* Login, using only their email address (no passwords), or using OAuth2/OIDC
* Select subscription and subscribe
* Manage their subscriptions, including unsubscribing
* Manage their customer account
  * List of subscriptions
  * List of invoices
  * Change street address for invoice
  * Change name
  * Change email addresses

### Customer support
* Our support staff needs to manage the customers. Similar to customer self-management above, but on the customers' behalf.

**Reply by minhchu** [reply]
@benbucksch I really appreciate your comment. 

Regarding the login functionality
> Login, using only their email address (no passwords), or using OAuth2/OIDC

 Do you override the default Medusa auth route, or do you create a new one to accommodate this functionality?

**Reply by benbucksch** [reply]
We don't have the Medusa shop implemented yet.

**Reply by Krislunde** [reply]
> Login, using only their email address (no passwords)

By this, I assume you mean the "we'll email you a login link for you to login without password" kind of login? Seems a little out of scope functionality-wise for a subscription plugin/module?

**Reply by benbucksch** [reply]
Agreed, login is out of scope for a subscription module. The subscription management UI would obviously also be an optional package on top of the core subscription logic. I added those requirements only for completeness.

**Comment by SalahAdDin** [comment]
Is there any ETA about it?

**Reply by manishkumar212111** [reply]
is there any ETA on it ??

**Reply by benbucksch** [reply]
@SalahAdDin, @manishkumar212111 Who are you asking?

**Reply by manishkumar212111** [reply]
> @SalahAdDin, @manishkumar212111 Who are you asking?

@klauss194

**Reply by manishkumar212111** [reply]
I have developed custom flow for subscription based order. It includes 

1. Digital wallet 
2. Order subscription
3. draft order addition
4. after confirmation fromadmin/store amount deducted from wallet and order created

**Reply by maximbroos** [reply]
@manishkumar212111 Can you share bits of your code? Your work sounds like a good foundation to start building this functionality.

**Comment by chrismarut** [comment]
Hey everyone! 👋
I'm Chris, co-founder of Reorder — a subscription engine built for Medusa as a plugin.
We've been working on solving the hard parts of recurring commerce inside Medusa: renewals, failed payment recovery (dunning), cancellation & retention workflows, and subscription analytics. All modeled as clean, separated domains within a single plugin architecture.
We just published a full technical demo walking through the entire system — real admin workflows, real customer scenarios, and the architectural decisions behind them:
https://www.youtube.com/watch?v=KY43_6Q3560&t=2s

If you're curious about what's under the hood:
📦 npm: https://www.npmjs.com/package/@reorderjs/reorder
💻 GitHub: https://github.com/reorder-js
🌐 Website: https://reorderjs.com/

The repo is public and the plugin is installable via npm, so feel free to dig in, use it as a starting point, or just study the patterns if you're building anything subscription-related on Medusa.
We'd love to hear your thoughts — and if you have any questions about the implementation, happy to chat here! 🙌
